File: clojure/ch01/ex1.8.clj
; Newton's Method for Cubes, using the approach in 1.7.
(ns sicp.ch01 (:use clojure.test))

(defn cube [x] (* x x x))

(defn improve [guess x]
  (/ (+ (/ x (* guess guess)) (* 2 guess)) 3))
  
(defn abs [x]
  (if (> x 0) x (- x)))

(defn good-enough? [guess prev-guess x]
  (< (abs (- guess prev-guess)) (* guess 0.00001)))
  
(defn cube-root-iter [guess prev-guess x]
  (if (good-enough? guess prev-guess x)
    guess
    (cube-root-iter (improve guess x) guess x)))

(defn cube-root [x]
  (cube-root-iter 1.0 0.0 x))

(deftest test-cube-root
  (is (< (abs (- (cube-root  8E15) 2E5)) 1.0))
  (is (< (abs (- (cube-root  8E-15) 2E-5)) 1.0)))

(run-tests)

File: scala/ch01/ex1.8.scala
// Newton's Method for Cubes, using the approach in 1.7.

def cube (x: Double) = x * x * x

def abs (x: Double) = math.abs(x)
  
def improve (guess: Double, x: Double) = 
  ((x / (guess * guess)) + (2.0 * guess)) / 3.0
  
def goodEnough (guess: Double, prevGuess: Double, x: Double) = 
  abs(guess - prevGuess) < (guess * 0.00001)

def cubeRootIter (guess: Double, prevGuess: Double, x: Double): Double =
  if (goodEnough(guess, prevGuess, x)) guess else cubeRootIter(improve(guess, x), guess, x)

def cubeRoot (x: Double) = cubeRootIter(1.0, 0.0, x)

import org.scalatest._
import org.scalatest.matchers._

object cubeRootSpec extends Spec with ShouldMatchers {
  describe ("Cube root calculator") {
    it ("should compute cube roots") {
      cubeRoot(8E15)  should be (2E5 plusOrMinus 1.0)
      cubeRoot(8E-15) should be (2E-5 plusOrMinus 1.0)
    }
  }
}
cubeRootSpec execute
